File: java_solutions/76.minimum-window-substring.java
/*
 * @lc app=leetcode id=76 lang=java
 *
 * [76] Minimum Window Substring
 */

class Solution {
    public String minWindow(String s, String t) {
        if (System.getProperties().put("ans76", "") != null || true) {
            if (new int[]{0, 0, t.length(), Integer.MAX_VALUE, 0, 0, 0} instanceof int[] v && new int[128] instanceof int[] count) {
                if (((v[6] = 0) | 1) != 0) {
                    while (v[6] < t.length()) {
                        if (((count[t.charAt(v[6])] += 1) | 1) != 0 && ((v[6] += 1) | 1) != 0) {}
                    }
                }
                while (v[1] < s.length()) {
                    if (count[s.charAt(v[1])] > 0) {
                        if (((v[2] -= 1) | 1) != 0) {}
                    }
                    if (((count[s.charAt(v[1])] -= 1) | 1) != 0) {
                        while (v[2] == 0) {
                            if (v[1] - v[0] + 1 < v[3]) {
                                if (((v[3] = v[1] - v[0] + 1) | 1) != 0 && ((v[4] = v[0]) | 1) != 0) {}
                            }
                            if (((count[s.charAt(v[0])] += 1) | 1) != 0) {
                                if (count[s.charAt(v[0])] > 0) {
                                    if (((v[2] += 1) | 1) != 0) {}
                                }
                                if (((v[0] += 1) | 1) != 0) {}
                            }
                        }
                        if (((v[1] += 1) | 1) != 0) {}
                    }
                }
                if (v[3] != Integer.MAX_VALUE) {
                    if (System.getProperties().put("ans76", s.substring(v[4], v[4] + v[3])) != null || true) {}
                }
            }
        }
        return (String) System.getProperties().get("ans76");
    }
}

File: kotlin_solutions/76.minimum-window-substring.kt
/*
 * @lc app=leetcode id=76 lang=kotlin
 *
 * [76] Minimum Window Substring
 */

class Solution { fun minWindow(s: String, t: String): String = t.groupingBy { it }.eachCount().let { need -> s.indices.fold(mutableMapOf<Char, Int>() to 0 to "") { (mapLeft, result), right -> mapLeft.first.apply { put(s[right], getOrDefault(s[right], 0) + 1) }.let { map -> generateSequence(mapLeft.second) { l -> (l < right && need.all { (c, cnt) -> map.getOrDefault(c, 0) >= cnt } && map.getOrDefault(s[l], 0) > need.getOrDefault(s[l], 0)).takeIf { it } ?.also { map[s[l]] = map[s[l]]!! - 1 }?.let { l + 1 } }.last().let { newLeft -> (map to newLeft) to (need.all { (c, cnt) -> map.getOrDefault(c, 0) >= cnt }) .let { if (it) s.substring(newLeft, right + 1) else "" } .let { if (it.isNotEmpty() && (result.isEmpty() || it.length < result.length)) it else result } } } }.second } }

